File: src/c/bounds_check.c
// 3️⃣ Bounds Checking and Memory Safety
// 
// Compile: gcc -O2 -S src/c/bounds_check.c -o target/bounds_check.s
// Assembly file: target/bounds_check.s
//
// Key assembly differences to observe:
// - C has no bounds checking, performs direct memory access
// - Assembly shows only simple memory address calculation
// - No safety check branches or validation code
// - May access undefined memory regions
// - Undefined behavior - may crash or return garbage values
// - Vulnerable to buffer overflow attacks
// - No runtime safety guarantees

#include <stdio.h>

int main() {
    int arr[3] = {1, 2, 3};
    // No bounds checking - direct access to out-of-bounds memory
    printf("%d\n", arr[5]); // Undefined behavior - may crash or return garbage
    return 0;
} 

File: src/rust/bounds_check.rs
// 3️⃣ Bounds Checking and Memory Safety
// 
// Compile: cargo rustc --bin bounds_check --release -- --emit=asm
// Assembly file: target/release/deps/bounds_check-*.s
//
// Key assembly differences to observe:
// - Rust inserts bounds checking code before array access
// - Out-of-bounds access triggers panic handler function call
// - Assembly shows comparison instructions checking if index exceeds bounds
// - Contains conditional jump instructions for panic branches
// - Release mode may optimize some bounds checks but not this example
// - Guaranteed memory safety at runtime with zero-cost when bounds are known
// - Prevents buffer overflow attacks and undefined behavior

use std::env;

fn main() {
    let arr = [1, 2, 3];
    // Use dynamic index to prevent compile-time checking
    let index = env::args().len() + 4; // This produces 5
    // This triggers panic - assembly shows bounds check and panic call
    println!("{}", arr[index]); // Index out of bounds, triggers panic
} 